Add Eviction Moratorium banner to the Tenant Platform

sraby opened the pull request:
This PR adds a banner announcing JustFix's status during the COVID-19 pandemic and links to RTC's FAQ page on the Eviction Moratorium.

I essentially created a front-end component that gets added to the `AppWithoutRouter` component— essentially rendering the banner below the nav bar on any page. I also included a "close" button that removes the banner, and it seems that the state of the banner stays consistent throughout at user session, so users only need to close the banner once to keep it hidden for their whole session on the Tenant Platform.

@toolness we're trying to push this out soon so lmk if you see any issues!

toolness commented on frontend/lib/moratorium-banner.tsx:
```diff
@@ -0,0 +1,43 @@
+  
+    }
+  
+    closeBanner = () => this.setState({isHidden: true});
```
Er, also, given the thing's propensity to potentially re-pop-up even after a user has dismissed it, I'm wondering if, in addition to only putting it on the few pages you suggested, we _also_ remove the close button: my hypothesis is that an un-closeable alert can potentially be less annoying than a recalcitrant alert that keeps showing up over and over.

That said, yet _another_ option is to store the state of the alert's display in the browser's `localStorage`.  On the vast majority of browsers this would mean that shoo-ing away the alert once would indeed shoo it away forever, much the way some sites have that annoying "IS IT OK IF WE USE COOKIES" alert that only shows up once.

Anyhow, if you do end up keeping the close button, I recommend additionally wrapping it in a `<SimpleProgressiveEnhancement>` component--this will ensure it's not shown to users in compatibility mode, which means they won't see a close box that is completely non-functional.

(Also if you keep the close button, I think this would be a great opportunity for you to use [React hooks](https://reactjs.org/docs/hooks-intro.html) instead of making a class-based component, but given all the other priorities on your plate I understand if you pass on that!)

sraby replied:
Thank you so much for all this thorough feedback. Always amazing to see how much you think about all the different kinds of users and user scenarios that may come up in our products! 

In terms of removing the "X" button, Tahnee was very keen on having one as for our mobile users, it was important to be able to collapse the banner as it tended to get very expansive on smaller devices. That being said, I think it totally makes sense to get rid of it on all pages and only keep it on the 5 pages we discussed at product check-in. When I noticed that the component's state stayed consistent throughout the lifetime of the web page, I think I got a little too excited and thought we could just keep it as is to catch any case where a user was visiting the Tenant Platform from an external page. But, given the nuisances you pointed out, it definitely seems like a good move to limit the presence of the banner except for where we really need it.

Will also work on removing the X button for users in compatibility mode and implementing it as a `<button>` for screen readers.

toolness replied:
Awesome, that sounds great to me! 🎉

toolness commented on frontend/sass/styles.scss:
```diff
@@ -323,3 +323,19 @@ html.jf-is-fullscreen-admin-page {
+            &:hover {
+                opacity: 0.7;
+            }
+        }
```
As I mentioned in an earlier comment, Bulma does have a `.delete` class and it'd be nice to use that if possible, both to maintain visual consistency and avoid CSS bloat, but it's up to you!

sraby replied:
👍

sraby commented on frontend/lib/moratorium-banner.tsx:
```diff
@@ -0,0 +1,43 @@
+        <section className={"jf-moratorium-banner hero is-warning is-small " + (this.state.isHidden && "is-hidden")}>
+          <div className="hero-body">
+            <div className="container">
+                <div className="jf-close-button is-size-5 is-pulled-right" onClick = {this.closeBanner}>✕</div>
```
yes, definitely! I can take a try at that

sraby commented:
Ok @toolness, I tried to address all of your concerns:

1. I restricted the presence of the banner to only our 4 main points of entry (root, loc splash, hp splash, rh splash)

2. I made the banner close button a progressive enhancement using your <SimpleProgressiveEnhancement> component

3. I refactored the SCSS to use the button delete class from Bulma, removing the need for a custom class there.

4. I refactored the MoratoriumBanner component to use react hooks! Love those!!

Let me know what you think!

toolness commented on frontend/lib/app.tsx:
```diff
@@ -30,6 +30,13 @@ import { areAnalyticsEnabled } from './analytics';
+  "/hp/splash",
+  "/rh/splash",
+  "/"
+];
```
This is cool for now since I doubt we'll internationalize the app before this moratorium is over, but just you know, if internationalization is enabled (which it isn't, on demo or production) these routes actually won't exist, since the locale-prefixed versions will exist instead (e.g. `/en/loc/splash`).  One way to future-proof it, as well as to avoid typos, is to use the `Routes` object/module rather than a hard-coded string, but there are some caveats around that that we probably don't want to deal with right now.  Anyhow just wanted to let you know for the future, I don't think you have to change anything here and now!

sraby replied:
Ah ok good to know! Hopefully the pandemic comes to an end before we have Tenant Platform internationalized...

toolness commented on frontend/lib/app.tsx:
```diff
@@ -305,7 +312,8 @@ export class AppWithoutRouter extends React.Component<AppPropsWithRouter, AppSta
                 <Navbar/>
-                <MoratoriumBanner />
+                {ROUTES_WITH_MORATORIUM_BANNER.includes(this.props.location.pathname) 
+                  && <MoratoriumBanner />}
```
Just so you know, one side-effect of conditionally displaying this component is that the state of the banner (open vs. closed) will be reset whenever it's unmounted and remounted... In effect that means that if I land on DDO and manually close the banner, it will re-appear when I click on one of the CTAs (since that's a splash page), which might be annoying.

To avoid this, you could _always_ render `<MoratoriumBanner>` but pass it `props.location.pathname` as a prop; then in _its_ render method/function, you could have it check to see if the passed-in pathname prop is one of the ones where it's actually supposed to appear, and if not, return `null`.  This means its open/closed state will be preserved for as long as the page lives, and the situation I described won't happen (at least not for JS-enabled users).

Anyways, it's just an enhancement, and I understand that this is pretty time-sensitive so it's definitely not a blocker!

toolness replied:
Oh wait!  Actually I don't think it will be unmounted/remounted when going from DDO to a splash page, since that won't result in an unmount/remount.  But if I e.g. click "get started" on the splash page but then click the back button--or if I go through the whole product flow and then click "take action" in the nav bar to go back to DDO--I think the banner would re-appear, which could be annoying.

sraby replied:
Interesting! I noticed that problem but didn't know of a workaround. I just implemented this and it seems to be working well on my end— lmk if it works for you too!

toolness commented on frontend/lib/moratorium-banner.tsx:
```diff
@@ -1,43 +1,29 @@
-    }
-  }
+    return (
+    <section className={"jf-moratorium-banner hero is-warning is-small " + (!isVisible && "is-hidden")}>
```
Er, I think that if `isVisible` is true, this section's `className` will be `"jf-moratorium-banner hero is-warning is-small false"`, which is probably unintentional.  I don't think this will actually have a detrimental effect since we probably don't have a CSS class called "false", but it's still a bit funky. 

Whenever I have a conditional class in addition to other classes, I use the [`classnames`](https://www.npmjs.com/package/classnames) package (which is already in our `package.json`) since it avoids these kinds of issues and usually makes the code a bit easier to read, but it's up to you.

sraby replied:
Oh right! I always forget about that package— so useful!

toolness commented on frontend/lib/moratorium-banner.tsx:
```diff
@@ -1,43 +1,29 @@
+                <span className="has-text-weight-bold">COVID-19 Update: </span>
+                JustFix.nyc is still in operation, and we are adapting our products to match new rules put in place during the Covid-19 public health crisis. 
+                Thanks to organizing from tenant leaders, renters now have stronger protections during this time, including a full halt on eviction cases. 
+                {' '}<a href="https://www.righttocounselnyc.org/moratorium_faq" rel="noopener noreferrer">
```
Did you also mean for this to have a `target="_blank"` on it?  Not sure if you want it to open in a new tab but that's what `rel="noopener"` usually accompanies.

sraby replied:
yep! Somehow that must have gotten missed.

sraby commented:
thank _you_!
